core/arch/aarch64/emit_utils.c

```diff
@@ -408,18 +408,40 @@ insert_load_dr_tls_base(dcontext_t *dcontext, instrlist_t *ilist, instr_t *where
 void
 append_fcache_enter_prologue(dcontext_t *dcontext, instrlist_t *ilist, bool absolute)
 {
+#ifdef UNIX
+    instr_t *no_signals = INSTR_CREATE_label(dcontext);
+    /* save callee-saved reg in case we return for a signal */
+    APP(ilist,
+        XINST_CREATE_move(dcontext, opnd_create_reg(DR_REG_X1),
+                          opnd_create_reg(REG_DCXT)));
+#endif
     ASSERT_NOT_IMPLEMENTED(!absolute &&
                            !TEST(SELFPROT_DCONTEXT, dynamo_options.protect_mask));
+
     /* Grab gen routine's parameter dcontext and put it into REG_DCXT:
      * mov x(dxct), x0
      */
     APP(ilist,
         XINST_CREATE_move(dcontext, opnd_create_reg(REG_DCXT),
                           opnd_create_reg(DR_REG_X0)));
 
-    /* FIXME i#2019: check dcontext->signals_pending.  First we need a way to
-     * create a conditional branch b.le.
-     */
+#ifdef UNIX
+    APP(ilist,
+        XINST_CREATE_load_1byte(
+            dcontext, opnd_create_reg(DR_REG_W2),
+            OPND_DC_FIELD(absolute, dcontext, OPSZ_1, SIGPENDING_OFFSET)));
+    APP(ilist,
+        XINST_CREATE_cmp(dcontext, opnd_create_reg(DR_REG_W2), OPND_CREATE_INT8(0)));
+    APP(ilist,
+        XINST_CREATE_jump_cond(dcontext, DR_PRED_LE, opnd_create_instr(no_signals)));
+
+    /* restore callee-saved reg */
+    APP(ilist,
+        XINST_CREATE_move(dcontext, opnd_create_reg(REG_DCXT),
+                          opnd_create_reg(DR_REG_X1)));
+    APP(ilist, XINST_CREATE_return(dcontext));
+    APP(ilist, no_signals);
+#endif
```

Comment:
Actually if I use DR_REG_R2/X2 instead of DR_REG_W2, encoding fails for the load.
```
Internal Error: Failed to encode instruction: 'ldrb   +0x0384(%x5)[1byte] -> %x2'
```
I think it needs a 32-bit width reg.

Reply:
But sure, I can move the logic to aarchxx, with some pre-processor directives to handle the differences between ARM and AARCH64.

Reply:
Done.